R/MeasureDensSquaredloss.R

```diff
@@ -0,0 +1,111 @@
+      # change `c("dens.kde")` to list of compatible learners
+      data = task$truth()
+
+      #calling the bandwidth
+      bw = switch(learner$id,
+             "dens.kde" = learner$param_set$values$bandwidth,
+             "dens.kde_ks" = learner$param_set$values$h,
+             "dens.nonpar" = learner$param_set$values$h,
+             "dens.kde_kd" = learner$param_set$values$bw,
+             "dens.mixed" = learner$param_set$values$bws
+             )
+
+      #vectorized the data
+      dat <- sapply(data, function (x, y) ((x - y) / bw), y = data)
+
+      if (learner$id == "dens.kde") {
+        kernel = ifelse(is.null(learner$param_set$values$kernel), "Epan", learner$param_set$values$kernel)
+      } else if (learner$id %in% c("dens.kde_ks", "dens.nonpar")) {
+        kernel = "Norm"
+      } else if (learner$id == "dens.mixed") {
+        kernel =  ifelse(is.null(learner$param_set$values$ckertype), "gaussian",
+                         learner$param_set$values$ckertype)
+      } else if (learner$id == "dens.kde_kd") {
+        kernel =  ifelse(is.null(learner$param_set$values$type_kernel), "n",
+                         learner$param_set$values$type_kernel)
+      }
+
+      if (kernel %in% c("e", "epan", "Epan", "epanechnikov")) {
+        squarednorm = distr6::Epanechnikov$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("n", "gaussian", "Norm", "NormalKernel")) {
+        squarednorm = distr6::NormalKernel$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("t", "Tri", "TriangularKernel")) {
+        squarednorm = distr6::TriangularKernel$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Sigm", "Sigmoid")) {
+        squarednorm = distr6::NormalKernel$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Quartic", "Quart", "b")) {
+        squarednorm = distr6::Quartic$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("cos", "Cosine")) {
+        squarednorm = distr6::Cosine$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("logis", "Logistic")) {
+        squarednorm = distr6::LogisticKernel$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Silv", "Silverman")) {
+        squarednorm = distr6::Silverman$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Tric", "Tricube")) {
+        squarednorm = distr6::Tricube$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Triw", "Triweight", "t")) {
+        squarednorm = distr6::Triweight$new()$pdfSquared2Norm(x = dat)
+      } else if (kernel %in% c("Unif", "UniformKernel", "uniform")) {
+        squarednorm = distr6::UniformKernel$new()$pdfSquared2Norm(x = dat)
+      } else {
+        squarednorm = NA
+      }
+
+      #squarednorm =  distr6::UniformKernel$new()$pdf(0.1)
+
+      pdf = prediction$pdf
+
+      if (is.null(bw)) {
+        score = NA
+      } else {
+        score = -2 * mean(pdf) + mean(squarednorm)/(length(data)^2 * bw)
+      }
+
```

Comment:
... something along the lines of
```
pdfeval = prediction$pdf(data)
pdfsquarednorm = prediction$intpdfsquared(-Inf, Inf)

score = -2 * mean(pdfeval) + mean(pdfsquarednorm)

Reply:
> pdfsquarednorm = prediction$intpdfsquared(-Inf, Inf)

This conflates the predicted distribution which internally uses a Kernel with the Kernel itself.

Reply:
I don´t quite get that - would the loss not only see the distribution and the distribution sees the kernel, but the loss doesn´t directly see the kernel?

Reply:
> I don´t quite get that - would the loss not only see the distribution and the distribution sees the kernel, but the loss doesn´t directly see the kernel?

Exactly. The loss only sees the fitted distribution, or the predicted pdf/cdf directly. However the distribution is essentially a black-box, the kernel is called within the pdf/cdf and is not a method of the distribution itself and nor is it a composite that wraps a kernel (we could consider this)